### PR TITLE
Feat/testing kpi staff controller

### PR DESCRIPTION
**Pull request by NaufalMuh11**
# Description

This pull request includes the following changes:

1. Addition of Database Transaction for Each Test:
     - Isolates new data created during testing to ensure a clean state for each test run.

2. Seeder Changes:
     - Updated the seeder to use `\database\db_new_api_weekly_report.sql`
     - Instructions to run the seeder after migration: php artisan db:seed.
     
3. Testing for Controllers:
    - Added tests for DivisionKpiController.
    - Added tests for FeedbackController.
    - Added tests for StaffKpiController.
 
Type of Change
- [ ]  New feature (non-breaking change which adds functionality)

## Additional Notes:

- Ensure to run `php artisan migrate` before seeding the database.
- To seed the database, run: `php artisan db:seed`
- Run test with `php artisan test`

### CODE REVIEW

**Review by nugrahaAdhitama — CHANGES_REQUESTED** [review]
# Code Review: `tests/Feature/DivisionKPITest.php`

## General Observations
- The tests cover various critical paths, including authorization, validation, and business rules, ensuring a well-rounded suite.
- The usage of `beforeEach` and `afterEach` hooks to manage database transactions is a great way to ensure tests are isolated.
- Assertions are consistent and clearly written, making the tests easy to follow.
- Good usage of Laravel’s response status codes, particularly for handling 403 and 422 errors.

## Specific Feedback

### **Scope 1: Duplication of Authorization Checks**
- **Issue**: Authorization checks are repeated across multiple test cases (e.g., checking if the user is unauthorized as a non-supervisor or clevel).
- **Suggestion**: Consider creating helper methods to avoid code duplication and improve test readability. For example, a method like `assertForbiddenForNonSupervisor()` could be reused across tests.

### **Scope 2: Hard-coded Email and Password Values**
- **Issue**: The hard-coded credentials (e.g., `ward.ruecker@example.com`, `'rahasia'`) reduce flexibility and scalability in the long term, and make the tests harder to maintain.
- **Suggestion**: Use factories or predefined constants to store test users and their credentials. This will make the code easier to update if credentials change or new users are added.

### **Scope 3: Lack of Data Validation Beyond Simple Errors**
- **Issue**: The tests validate simple errors like forbidden actions or incorrect data types but don’t seem to cover edge cases like missing fields or boundary values (e.g., negative weights).
- **Suggestion**: Add more test cases for edge scenarios, such as weights exceeding 100 or missing KPI fields. This will make your test coverage more comprehensive.

### **Scope 4: Use of `expect` Instead of Laravel Assertions**
- **Issue**: You're using `expect` for your assertions, which, while readable, doesn’t fully leverage Laravel’s built-in testing features, such as `assertForbidden()` or `assertJsonValidationErrors()`.
- **Suggestion**: Consider using Laravel’s built-in testing methods for better readability and adherence to Laravel conventions. For instance, replace `expect($response->status())->toBe(403)` with `$response->assertForbidden()`.

### **Scope 5: Insufficient Test Coverage for GET Requests**
- **Issue**: The tests for GET requests focus heavily on error scenarios like unauthorized access or data not found. There’s less emphasis on testing valid responses.
- **Suggestion**: Include more positive test cases for valid GET requests, including different months, years, or scenarios where partial data may exist.

## Conclusion
Overall, the tests achieve their goal of covering core API paths for the KPI feature. However, there are several areas where improvements can be made to enhance maintainability and scalability. Introducing helper methods to reduce code duplication, using Laravel’s built-in assertion methods, and expanding test coverage to include edge cases will help you align better with clean code principles and best practices in unit testing. These changes will also ensure that the tests remain readable, maintainable, and effective as the system evolves. Keep up the good work!

File: tests/Feature/DivisionKPITest.php [hunk 1]
```diff
@@ -1,26 +1,24 @@
...
+});
+
+afterEach(function () {
+    DB::rollBack();
+});
 
 describe('POST /api/v1/kpi/supervisor-division/{year}/{month}', function () {
     it('returns 403 unauthorized if user is not supervisor', function () {
-            $staffToken = authenticateAs('turner.emmet@example.org', 'rahasia');
+        $staffToken = authenticateAs('turner.emmet@example.org', 'rahasia');
```
Comment (nugrahaAdhitama) [inline tests/Feature/DivisionKPITest.php:15]: This is just used as a variable

Comment (nugrahaAdhitama) [inline tests/Feature/DivisionKPITest.php:15]: This is just used as a variable

File: tests/Feature/DivisionKPITest.php [hunk 2]
```diff
@@ -1,26 +1,24 @@
...
 
-            expect($response->status())->toBe(403);
-            expect($response->json())->toMatchArray(['message' => 'Forbidden']);
-        });
+        $response->assertForbidden()
+         ->assertJson(['message' => 'Forbidden']);
+    });
 
     it('allows a supervisor to submit KPIs for a division', function () {
         $supervisorToken = authenticateAs('ward.ruecker@example.com', 'rahasia');
```
Comment (nugrahaAdhitama) [inline tests/Feature/DivisionKPITest.php:24]: This is also used as a variable

File: tests/Feature/DivisionKPITest.php [hunk 3]
```diff
@@ -115,14 +107,12 @@
...
-        expect($response->json())->toEqual([
-            'message' => 'Total weight must be 100',
-        ]);
+        $response->assertStatus(Response::HTTP_LOCKED)
+         ->assertJson(['message' => 'Total weight must be 100']);
     });
 
     it('returns 403 with error when KPI already exists', function () {
-         $supervisorToken = authenticateAs('ward.ruecker@example.com', 'rahasia');
+        $supervisorToken = authenticateAs('ward.ruecker@example.com', 'rahasia');
```
Comment (nugrahaAdhitama) [inline tests/Feature/DivisionKPITest.php:115]: This is also used as a variable

File: tests/Feature/DivisionKPITest.php [hunk 4]
```diff
@@ -134,43 +124,42 @@
...
+         ->assertJson([
             'message' => 'KPIs for the given division, year, and month already exist and cannot be modified',
         ]);
     });
 });
 
 describe('GET /api/v1/kpi/supervisor-division/{year}/{month}', function () {
-     it('returns 403 if user is not supervisor', function () {
+    it('returns 403 if user is not supervisor', function () {
         $staffToken = authenticateAs('turner.emmet@example.org', 'rahasia');
```
Comment (nugrahaAdhitama) [inline tests/Feature/DivisionKPITest.php:140]: This is also used as a variable

File: tests/Feature/DivisionKPITest.php [hunk 5]
```diff
@@ -134,43 +124,42 @@
...
 
-        expect($response->status())->toBe(403);
-        expect($response->json())->toMatchArray(['message' => 'Forbidden']);
+        $response->assertForbidden()
+         ->assertJson(['message' => 'Forbidden']);
     });
 
-    it('returns 404 if KPI data is not found', function(){
+    it('returns 404 if KPI data is not found', function () {
         $supervisorToken = authenticateAs('ward.ruecker@example.com', 'rahasia');
```
Comment (nugrahaAdhitama) [inline tests/Feature/DivisionKPITest.php:149]: This is also used as a variable

File: tests/Feature/DivisionKPITest.php [hunk 6]
```diff
@@ -134,43 +124,42 @@
...
         $response = $this->withToken($supervisorToken)->getJson('/api/v1/kpi/supervisor-division/2024/5');
 
-        expect($response->status())->toBe(404);
-        expect($response->json())->toMatchArray(['message' => "Data not found."]);
+        $response->assertNotFound()
+         ->assertJson(['message' => 'Data not found.']);
     });
 
     it('returns 200 with KPI data if authenticated and authorized', function () {
         $supervisorToken = authenticateAs('ward.ruecker@example.com', 'rahasia');
```
Comment (nugrahaAdhitama) [inline tests/Feature/DivisionKPITest.php:157]: This is also used as a variable

File: tests/Feature/DivisionKPITest.php [hunk 7]
```diff
@@ -198,38 +187,40 @@
 
 describe('POST api/v1/kpi/clevel/{divisionId}/{year}/{month}/score', function () {
     it('returns 403 unauthorized if user is not clevel', function () {
-         $supervisorToken = authenticateAs('ward.ruecker@example.com', 'rahasia');
+        $supervisorToken = authenticateAs('ward.ruecker@example.com', 'rahasia');
```
Comment (nugrahaAdhitama) [inline tests/Feature/DivisionKPITest.php:190]: This is also used as a variable

File: tests/Feature/DivisionKPITest.php [hunk 8]
```diff
@@ -198,38 +187,40 @@
...
 
-        expect($response->status())->toBe(403);
-        expect($response->json())->toMatchArray(['message' => 'Forbidden']);
+        $response->assertForbidden()
+         ->assertJson(['message' => 'Forbidden']);
     });
+
     it('returns 403 if clevel division not valid', function () {
-         $supervisorToken = authenticateAs('karolann97@example.com', 'rahasia');
+        $supervisorToken = authenticateAs('karolann97@example.com', 'rahasia');
```
Comment (nugrahaAdhitama) [inline tests/Feature/DivisionKPITest.php:201]: This is also used as a variable

File: tests/Feature/DivisionKPITest.php [hunk 9]
```diff
@@ -264,63 +256,58 @@
...
-                        'realizations' => ['The realizations field is required.']
-                    ],
-                ]);
+        $response->assertUnprocessable()
+            ->assertJsonValidationErrors(['realizations']);
     });
 
     it('returns 422 if number of realizations does not match KPIs', function () {
-       $supervisorToken = authenticateAs('josue60@example.com', 'rahasia');
+        $supervisorToken = authenticateAs('josue60@example.com', 'rahasia');
```
Comment (nugrahaAdhitama) [inline tests/Feature/DivisionKPITest.php:264]: This is also used as a variable

File: tests/Feature/DivisionKPITest.php [hunk 10]
```diff
@@ -264,63 +256,58 @@
...
 
-        $response->assertStatus(422)
-                ->assertJson(['message' => 'Number of realizations must match the number of KPIs']);
+        $response->assertUnprocessable()
+            ->assertJson(['message' => 'Number of realizations must match the number of KPIs']);
     });
 
     it('returns 422 if realization exceeds target', function () {
-         $supervisorToken = authenticateAs('josue60@example.com', 'rahasia');
+        $supervisorToken = authenticateAs('josue60@example.com', 'rahasia');
```
Comment (nugrahaAdhitama) [inline tests/Feature/DivisionKPITest.php:275]: This is also used as a variable

File: tests/Feature/DivisionKPITest.php [hunk 11]
```diff
@@ -264,63 +256,58 @@
...
+        $response->assertUnprocessable()
+            ->assertJson(['message' => 'End-of-month realization cannot exceed the target']);
     });
 });
 
 describe('GET api/v1/kpi/clevel/{divisionId}/{year}/{month}/score', function () {
-
     it('returns 403 if user is not clevel', function () {
-         $supervisorToken = authenticateAs('ward.ruecker@example.com', 'rahasia');
+        $supervisorToken = authenticateAs('ward.ruecker@example.com', 'rahasia');
```
Comment (nugrahaAdhitama) [inline tests/Feature/DivisionKPITest.php:288]: This is also used as a variable

File: tests/Feature/DivisionKPITest.php [hunk 12]
```diff
@@ -264,63 +256,58 @@
...
 
-        expect($response->status())->toBe(403);
-        expect($response->json())->toMatchArray(['message' => 'Forbidden']);
+        $response->assertForbidden()
+         ->assertJson(['message' => 'Forbidden']);
     });
 
     it('returns 403 if clevel division not valid', function () {
-         $supervisorToken = authenticateAs('karolann97@example.com', 'rahasia');
+        $supervisorToken = authenticateAs('karolann97@example.com', 'rahasia');
```
Comment (nugrahaAdhitama) [inline tests/Feature/DivisionKPITest.php:297]: This is also used as a variable

File: tests/Feature/DivisionKPITest.php [hunk 13]
```diff
@@ -264,63 +256,58 @@
...
 
-        expect($response->status())->toBe(403);
-        expect($response->json())->toMatchArray(['message' => 'You are not authorized to see this division']);
+        $response->assertForbidden()
+         ->assertJson(['message' => 'You are not authorized to see this division']);
     });
 
-    it('returns 404 if KPI data is not found', function(){
+    it('returns 404 if KPI data is not found', function () {
         $supervisorToken = authenticateAs('josue60@example.com', 'rahasia');
```
Comment (nugrahaAdhitama) [inline tests/Feature/DivisionKPITest.php:306]: This is also used as a variable

**Review by nugrahaAdhitama — CHANGES_REQUESTED** [review]
For emails and passwords in all files, it's better to put them in a variable, so that if there are changes in the future. Just change the contents of the variable and don't need to search for each one.

**Comment by NaufalMuh11** [comment]
Done 🙏

**Review by nugrahaAdhitama — APPROVED** [review]
Approved
